doc/developer/design/20210831_correctness.md

```diff
@@ -0,0 +1,124 @@
+A definite collection has had its updates put in a specific order, by way of the timestamps on updates.
+A handle to a collection can be "definite from `since` until `upper`" (names pending) if it reflects the definite results for times greater or equal to `since` but not greater or equal to `upper`.
+Among the correctness guarantees of Materialize, a main one will be statements of definiteness over intervals of times for collections it maintains.
+
+### Components
+
+There are three components to discuss: storage, compute, and coordination.
+
+The **storage layer** (`persist`, plus potentially others) presents streams of updates to collections to other parts of the system (primary the compute layer).
+One correctness goal of the `persist` module is to make as many update streams as possible "definite", even if the raw data are indefinite (e.g. perhaps they contain no timestamps, which must be added).
+The storage layer does not need to capture the entire contents of a collection to make it definite; durable "offset"-to-timestamp bindings can suffice if the source data are durable but without timestamps, and no data may be required if the source data are durable with timestamps (e.g. MzCDC sources).
+Update streams that are definite form the "durable storage" of Materialize, and are the basis for Materialize's durability guarantees.
+
+The **compute layer** (`dataflow`, plus potentially others) consumes streams of updates to collections, and computes and then incrementally maintains the results of views over these data.
+A view is ideally "deterministic" in the sense that it acts as a function on the input update streams, and produces a well-defined output on definite input (we can allow "non-deterministic" views but like "indefinite streams" they taint their results).
+One correctness goal of the `dataflow` module is to convert definite inputs into definite outputs, which can be returned to the storage layer or to interactive user sessions.
+Views and their dependencies form the "concurrency control" of Materialize, in that outputs are reported as definite only when, and ideally as soon as, their inputs are definite.
+
+The **coordination layer** (`coord`, plus potentially others) interfaces with user sessions and pilots the storage and compute layers.
+It maintains mappings from user names to storage and compute objects (source, views, indexes), and dereferences them in ways that provide a consistent experience for users.
+It also chooses timestamps for reads (peeks) and writes (table inserts) to provide the lived experience of continually moving forward through time.
+It pilots the storage and compute layers and is responsible for ensuring they communicate clearly and completely with each other.
+One correctness goal of the `coord` module is to provide "strict sequential consistency" to the interactive users, through the careful use of the storage and compute layers.
```

Comment:
A formal definition of "strict sequential consistency" could be useful here

Reply:
Is it ok to link to one? The consistency levels are famously quirky, and I'll likely botch it. If it's about making sure that a reader can get to a definition, that is easier than making this self-contained. I'll plan to link one, but lmk if you meant something else.

I'm also comfortable with "One correctness goal ... is to pick and provide a SQL consistency level ..."; I just happen to think we have one in mind and referenced it by name.

Reply:
This is the link I would use; lmk if it looks good, if you have a chance (just a scan for whether it seems likely to provide the clarity that just using the name above did not): https://jepsen.io/consistency/models/strict-serializable

I seem to have screwed up the name is part of the problem; it seems it should be either "strict consistency" or "strict serializable consistency", even though "strict consistency" improves on "sequential consistency" (https://en.wikipedia.org/wiki/Sequential_consistency). Sorry.

Reply:
Yep, the Jepsen link looks great.